src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -148,38 +162,25 @@
     @Override
     public void beginTree(DetailAST rootAST) {
         classes = new ArrayDeque<>();
+        innerClasses = new TreeMap<>();
         packageName = "";
     }
 
     @Override
     public void visitToken(DetailAST ast) {
-        final DetailAST modifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
-
         switch (ast.getType()) {
             case TokenTypes.PACKAGE_DEF:
                 packageName = extractQualifiedName(ast.getFirstChild().getNextSibling());
                 break;
 
             case TokenTypes.CLASS_DEF:
-                registerNestedSubclassToOuterSuperClasses(ast);
-
-                final boolean isFinal = modifiers.findFirstToken(TokenTypes.FINAL) != null;
-                final boolean isAbstract = modifiers.findFirstToken(TokenTypes.ABSTRACT) != null;
-
-                final String qualifiedClassName = getQualifiedClassName(ast);
-                classes.push(new ClassDesc(qualifiedClassName, isFinal, isAbstract));
+            case TokenTypes.ENUM_DEF:
+            case TokenTypes.RECORD_DEF:
+                visitType(ast);
                 break;
 
             case TokenTypes.CTOR_DEF:
-                if (!ScopeUtil.isInEnumBlock(ast) && !ScopeUtil.isInRecordBlock(ast)) {
-                    final ClassDesc desc = classes.peek();
-                    if (modifiers.findFirstToken(TokenTypes.LITERAL_PRIVATE) == null) {
-                        desc.registerNonPrivateCtor();
-                    }
-                    else {
-                        desc.registerPrivateCtor();
-                    }
-                }
+                visitCtor(ast);
```

Comment:
Extracted to reduce method complexity.

Reply:
Is it possible to make this PR into two separate commits, and have the first one be something like `minor: refactor FinalClassCheck to ease update for #10859`, then have a commit for functional changes that directly relate to #10859 ?

Reply:
@pbludov ping

Reply:
Sorry for the delay. I'll split it.

Reply:
done.